File: m3-h5/src/test/java/ru/yegorpilipenko/otus/spring/m2h4/controller/BookControllerTest.java
package ru.yegorpilipenko.otus.spring.m2h4.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.val;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.hateoas.mvc.ResourceAssemblerSupport;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.web.servlet.MockMvc;
import ru.yegorpilipenko.otus.spring.m2h4.dto.ShortBook;
import ru.yegorpilipenko.otus.spring.m2h4.model.Author;
import ru.yegorpilipenko.otus.spring.m2h4.model.Book;
import ru.yegorpilipenko.otus.spring.m2h4.model.Comment;
import ru.yegorpilipenko.otus.spring.m2h4.model.Genre;
import ru.yegorpilipenko.otus.spring.m2h4.resource.BookResource;
import ru.yegorpilipenko.otus.spring.m2h4.resource.ShortBookResource;
import ru.yegorpilipenko.otus.spring.m2h4.service.BookService;

import java.util.List;
import java.util.Optional;

import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_EMPTY;
import static org.hamcrest.Matchers.isEmptyString;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.springframework.hateoas.MediaTypes.HAL_JSON_UTF8;
import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static ru.yegorpilipenko.otus.spring.m2h4.controller.BookController.URL;

@WebMvcTest(BookController.class)
@ExtendWith(SpringExtension.class)
class BookControllerTest {

    @MockBean
    BookService bookService;
    @MockBean
    ResourceAssemblerSupport<Book, BookResource> bookResourceAssembler;
    @MockBean
    ResourceAssemblerSupport<ShortBook, ShortBookResource> shortBookResourceAssembler;
    @Autowired
    MockMvc mockMvc;
    ObjectMapper objectMapper = new ObjectMapper().setDefaultPropertyInclusion(NON_EMPTY);

    @AfterEach
    void clean() {
        // workaround for compatibility @MockBean and @Nested
        reset(bookService, bookResourceAssembler, shortBookResourceAssembler);
    }

    @Nested
    class GetAllTest {

        @Test
        void shouldPassForExistsBooks() throws Exception {
            val books = List.of(new ShortBookResource("n1"), new ShortBookResource("n2"));
            doReturn(null).when(bookService).getAll();
            doReturn(books).when(shortBookResourceAssembler).toResources(any());
            mockMvc.perform(get(URL))
                    .andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(HAL_JSON_UTF8))
                    .andExpect(content().json(objectMapper.writeValueAsString(books)));
            verify(bookService).getAll();
            verify(shortBookResourceAssembler).toResources(any());
            verifyNoMoreInteractions(bookService, shortBookResourceAssembler);
        }

        @Test
        void shouldPassForNotExistsBooks() throws Exception {
            doReturn(null).when(bookService).getAll();
            doReturn(List.of()).when(shortBookResourceAssembler).toResources(any());
            mockMvc.perform(get(URL))
                    .andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(HAL_JSON_UTF8))
                    .andExpect(content().json("[]"));
            verify(bookService).getAll();
            verify(shortBookResourceAssembler).toResources(any());
            verifyNoMoreInteractions(bookService, shortBookResourceAssembler);
        }

    }

    @Nested
    class GetByIdTest {

        @Test
        void shouldPassForExistsBook() throws Exception {
            val bookId = "1";
            val book = new BookResource(
                    "n",
                    List.of(new Author("f", "l")),
                    List.of(new Genre("gn")),
                    List.of(new Comment("e", "t"))
            );
            doReturn(Optional.of(Book.builder().build())).when(bookService).getById(bookId);
            doReturn(book).when(bookResourceAssembler).toResource(any());
            mockMvc.perform(get("{base}/{bookId}", URL, bookId))
                    .andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(HAL_JSON_UTF8))
                    .andExpect(content().json(objectMapper.writeValueAsString(book)));
            verify(bookService).getById(bookId);
            verify(bookResourceAssembler).toResource(any());
            verifyNoMoreInteractions(bookService, bookResourceAssembler);
        }

        @Test
        void shouldPassForNotExistsBook() throws Exception {
            val bookId = "1";
            doReturn(Optional.empty()).when(bookService).getById(bookId);
            mockMvc.perform(get("{base}/{bookId}", URL, bookId))
                    .andDo(print())
                    .andExpect(status().isNotFound())
                    .andExpect(content().string(isEmptyString()));
            verify(bookService).getById(bookId);
            verifyNoMoreInteractions(bookService, bookResourceAssembler);
        }

    }

    @Nested
    class SaveTest {

        @Test
        void shouldPass() throws Exception {
            val book = new BookResource(
                    "n",
                    List.of(new Author("f", "l")),
                    List.of(new Genre("gn")),
                    List.of(new Comment("e", "t"))
            );
            doReturn(null).when(bookService).save(any());
            doReturn(book).when(bookResourceAssembler).toResource(any());
            mockMvc.perform(
                    post(URL)
                            .contentType(APPLICATION_JSON_UTF8)
                            .content(objectMapper.writeValueAsString(book))
            ).andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(HAL_JSON_UTF8))
                    .andExpect(content().json(objectMapper.writeValueAsString(book)));
        }

    }

    @Nested
    class RemoveByIdTest {

        @Test
        void shouldPassForExistsBook() throws Exception {
            val bookId = "1";
            doReturn(true).when(bookService).removeById(bookId);
            mockMvc.perform(delete("{base}/{bookId}", URL, bookId))
                    .andDo(print())
                    .andExpect(status().isOk())
                    .andExpect(content().string(isEmptyString()));
            verify(bookService).removeById(bookId);
            verifyNoMoreInteractions(bookService);
        }

        @Test
        void shouldPassForNotExistsBook() throws Exception {
            val bookId = "1";
            doReturn(false).when(bookService).removeById(bookId);
            mockMvc.perform(delete("{base}/{bookId}", URL, bookId))
                    .andDo(print())
                    .andExpect(status().isNotFound())
                    .andExpect(content().string(isEmptyString()));
            verify(bookService).removeById(bookId);
            verifyNoMoreInteractions(bookService);
        }

    }

}

File: m1-h5/src/test/kotlin/ru/yegorpilipenko/otus/spring/m1h5/controller/BookControllerTest.kt
package ru.yegorpilipenko.otus.spring.m1h5.controller

import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.assertj.core.api.WithAssertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import ru.yegorpilipenko.otus.spring.m1h5.dto.ShortBook
import ru.yegorpilipenko.otus.spring.m1h5.model.Author
import ru.yegorpilipenko.otus.spring.m1h5.model.Book
import ru.yegorpilipenko.otus.spring.m1h5.model.Genre
import ru.yegorpilipenko.otus.spring.m1h5.service.BookService
import java.io.ByteArrayInputStream
import java.util.Scanner

@ExtendWith(MockitoExtension::class)
class BookControllerTest : WithAssertions {

    @Mock lateinit var bookService: BookService
    lateinit var scanner: Scanner
    lateinit var controller: BookController

    @BeforeEach
    fun setup() {
        scanner = Scanner(ByteArrayInputStream("""
            2 3
            4 5
        """.trimIndent().toByteArray()))
        controller = BookController(bookService, scanner)
    }

    @Nested
    inner class BooksTest {

        @Test
        fun shouldPassForMoreBooks() {
            doReturn(listOf(ShortBook(1, "b1"), ShortBook(2, "b2"))).whenever(bookService).findAll()
            assertThat(controller.books()).containsExactlyInAnyOrder("1: b1", "2: b2")
        }

        @Test
        fun shouldPassWithoutBooks() {
            doReturn(emptyList<ShortBook>()).whenever(bookService).findAll()
            assertThat(controller.books()).isEmpty()
        }
    }

    @Nested
    inner class BookTest {

        @Test
        fun shouldPassForExistsBook() {
            val id = 1L
            doReturn(Book(id, "b", listOf(Author(2, "af", "al")), listOf(Genre(3, "g"))))
                    .whenever(bookService).findById(id)
            assertThat(controller.book(id)).isEqualTo("""
                1: b
                Авторы:
                2: af al
                Жанры:
                3: g
            """.trimIndent())
        }

        @Test
        fun shouldPassForExistsShortBook() {
            val id = 1L
            doReturn(Book(id, "b", emptyList(), emptyList())).whenever(bookService).findById(id)
            assertThat(controller.book(id)).isEqualTo("1: b")
        }

        @Test
        fun shouldPassForNotExistsBook() {
            val id = 1L
            doReturn(null).whenever(bookService).findById(id)
            assertThat(controller.book(id)).isEqualTo("Not found")
        }
    }

    @Nested
    inner class BookSaveTest {

        @Test
        fun shouldPassForSaveFullBook() {
            doReturn(Book(
                    1,
                    "b",
                    listOf(Author(2, "af1", "al1"), Author(3, "af2", "al2")),
                    listOf(Genre(4, "g1"), Genre(5, "g2")))
            ).whenever(bookService).save(any(), any(), any())
            assertThat(controller.bookSave("b")).isEqualTo("""
                1: b
                Авторы:
                2: af1 al1
                3: af2 al2
                Жанры:
                4: g1
                5: g2
            """.trimIndent())
            verify(bookService).save(ShortBook(name = "b"), listOf(2L, 3L), listOf(4L, 5L))
        }
    }

    @Nested
    inner class BookRemoveTest {

        @Test
        fun shouldPassForExistsBook() {
            val id = 1L
            doReturn(true).whenever(bookService).removeById(id)
            assertThat(controller.bookRemove(id)).isEqualTo("Removed")
        }

        @Test
        fun shouldPassForNotExistsBook() {
            val id = 1L
            doReturn(false).whenever(bookService).removeById(id)
            assertThat(controller.bookRemove(id)).isEqualTo("Not found")
        }
    }
}
